### PR TITLE
Have i been Pwned fix

### PR DESCRIPTION
**Pull request by teizenman**
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/18102
fixes: https://github.com/demisto/etc/issues/18781

## Description
Removed 'Mandatory' tag from commands' arguments

### CODE REVIEW

File: Integrations/Pwned-V2/CHANGELOG.md [hunk 1]
```diff
@@ -1,4 +1,7 @@
 ## [Unreleased]
+    - Removed 'Mandatory' tag from commands' arguments
```
Comment (ronykoz) [inline Integrations/Pwned-V2/CHANGELOG.md]: Not relevant for customers

File: Integrations/Pwned-V2/CHANGELOG.md [hunk 2]
```diff
@@ -1,4 +1,7 @@
...
+    - Added support to rate-limit issue
```
Comment (ronykoz) [inline Integrations/Pwned-V2/CHANGELOG.md]: same here

File: Integrations/Pwned-V2/Pwned-V2.py [hunk 3]
```diff
@@ -48,7 +48,13 @@ def http_request(method, url_suffix, params=None, data=None):
 
     if res.status_code == 404:
         return None
-    elif not res.status_code == 200:
+    elif res.status_code == 429:
+        # Rate limit response code
+        wait_regex = re.search(r'\d+', res.json()['message'])
```
Comment (ronykoz) [inline Integrations/Pwned-V2/Pwned-V2.py]: are you sure that just searching for a number is enough?

File: Integrations/Pwned-V2/Pwned-V2.py [hunk 4]
```diff
@@ -155,11 +161,13 @@ def add_malicious_to_context(malicious_type):
     }
 
 
-def email_to_entry_context(email, hibp_email_res, hibp_paste_res):
+def email_to_entry_context(email, api_email_res, api_paste_res):
     dbot_score = 0
     comp_email = dict()  # type: dict
-    comp_sites = sorted([item['Title'] for item in hibp_email_res])
-    comp_pastes = sorted(set(item['Source'] for item in hibp_paste_res))
+    comp_sites = [item['Title'] for item in api_email_res]
```
Comment (ronykoz) [inline Integrations/Pwned-V2/Pwned-V2.py]: make them both one liners again

File: Integrations/Pwned-V2/Pwned-V2.py [hunk 5]
```diff
@@ -181,12 +189,25 @@ def domain_to_entry_context(domain, hibp_res):
...
+def rate_limit_retry(amount_of_seconds, request_type):
+    time.sleep(amount_of_seconds)
+    if request_type == 'email':
+        pwned_email_command()
+    else:
+        pwned_domain_command()
+
+
+def retry_needed(api_res, api_paste_res):
+    return api_res and api_paste_res and ('request_retry' in api_res or 'request_retry' in api_paste_res)
```
Comment (ronykoz) [inline Integrations/Pwned-V2/Pwned-V2.py]: split into two conditions, this usage of api_paste_res which doesn't always exists is weird

File: Integrations/Pwned-V2/Pwned-V2.py [hunk 6]
```diff
@@ -48,7 +48,13 @@ def http_request(method, url_suffix, params=None, data=None):
 
     if res.status_code == 404:
         return None
-    elif not res.status_code == 200:
+    elif res.status_code == 429:
+        # Rate limit response code
+        wait_regex = re.search(r'\d+', res.json()['message'])
+        if wait_regex:
+            wait_amount = wait_regex.group()
+            return {'request_retry': int(wait_amount)}
```
Comment (ronykoz) [inline Integrations/Pwned-V2/Pwned-V2.py]:
```suggestion
            return {'time_to_wait': int(wait_amount)}
```

**Comment by ronykoz** [comment]
@kirbles19 please review the CHANGELOG.md file please

**Comment by kirbles19** [comment]
@teizenman @ronykoz Done. Please check for technical accuracy.

File: Integrations/Pwned-V2/Pwned-V2.py [hunk 7]
```diff
@@ -187,6 +197,29 @@ def domain_to_entry_context(domain, hibp_res):
...
+def retry_needed(api_res, api_paste_res=None):
+    global retries_end_time
+    if retries_end_time == datetime.min and not MAX_RETRY_ALLOWED == -1:
+        retries_end_time = datetime.now() + timedelta(seconds=int(MAX_RETRY_ALLOWED))
+
+    if api_res and 'time_to_wait' in api_res:
+        return True
+    elif api_paste_res and 'time_to_wait' in api_paste_res:
+        return True
+    return False
```
Comment (yaakovi) [inline Integrations/Pwned-V2/Pwned-V2.py]: add an empty line before this one

File: Integrations/Pwned-V2/Pwned-V2.py [hunk 8]
```diff
@@ -187,6 +197,29 @@ def domain_to_entry_context(domain, hibp_res):
...
+    time.sleep(amount_of_seconds)
+    if request_type == 'email':
+        pwned_email_command()
+    else:
+        pwned_domain_command()
+
+
+def retry_needed(api_res, api_paste_res=None):
+    global retries_end_time
+    if retries_end_time == datetime.min and not MAX_RETRY_ALLOWED == -1:
```
Comment (yaakovi) [inline Integrations/Pwned-V2/Pwned-V2.py]:
```suggestion
    if retries_end_time == datetime.min and MAX_RETRY_ALLOWED != -1:
```

File: Integrations/Pwned-V2/Pwned-V2.yml [hunk 9]
```diff
@@ -7,24 +7,32 @@ configuration:
   name: api_key
   required: true
   type: 4
+- display: Maximum time per request (In seconds)
+  name: max_retry_time
+  required: false
+  type: 0
```
Comment (glicht) [inline Integrations/Pwned-V2/Pwned-V2.yml]: Add a default value of 30

File: Integrations/Pwned-V2/Pwned-V2.py [hunk 10]
```diff
@@ -33,22 +35,37 @@
...
+            # Rate limit response code
+            break
+
+        if datetime.now() > RETRIES_END_TIME:
+            return_error('Max retry time has exceeded.')
+
+        wait_regex = re.search(r'\d+', res.json()['message'])
+        if wait_regex:
+            wait_amount = wait_regex.group()
+            time.sleep(int(wait_amount))
```
Comment (glicht) [inline Integrations/Pwned-V2/Pwned-V2.py]: Calculate if wait is going to exceed RETRIES_END_TIME. If so, don't wait and return an error immediately.

**Comment by kirbles19** [comment]
@teizenman Done.

**Pull request merged**